independent-projects/qute/generator/src/main/java/io/quarkus/qute/generator/ExtensionMethodGenerator.java

```diff
@@ -175,490 +170,474 @@ public String generate(MethodInfo method, String matchName, List<String> matchNa
-
-        private final ClassCreator namespaceResolver;
-
-        public NamespaceResolverCreator(ClassInfo declaringClass, String namespace, int priority) {
-            String baseName;
-            if (declaringClass.enclosingClass() != null) {
-                baseName = simpleName(declaringClass.enclosingClass()) + ValueResolverGenerator.NESTED_SEPARATOR
-                        + simpleName(declaringClass);
+            if (!params.needsEvaluation()) {
+                Expr[] args = new Expr[params.size()];
+                for (int i = 0; i < params.size(); i++) {
+                    Param param = params.get(i);
+                    if (param.kind == ParamKind.NAME) {
+                        args[i] = name;
+                    } else if (param.kind == ParamKind.ATTR) {
+                        args[i] = nested.invokeInterface(Descriptors.GET_ATTRIBUTE, evalContext,
+                                Const.of(param.name));
+                    }
+                }
+                nested.return_(nested.invokeStatic(Descriptors.COMPLETED_STAGE_OF,
+                        nested.invokeStatic(methodDescOf(method), args)));
             } else {
-                baseName = simpleName(declaringClass);
+                LocalVar ret = nested.localVar("ret", nested.new_(CompletableFuture.class));
+                // Evaluate params first
+                // The CompletionStage upon which we invoke whenComplete()
+                LocalVar evaluatedParams = nested.localVar("evaluatedParams",
+                        nested.invokeStatic(Descriptors.EVALUATED_PARAMS_EVALUATE,
+                                evalContext));
+                Expr paramsReady = evaluatedParams.field(Descriptors.EVALUATED_PARAMS_STAGE);
+
+                // Function that is called when params are evaluated
+                Expr whenCompleteFun = nested.lambda(BiConsumer.class, lc -> {
+                    Var capturedName = isNameParamRequired ? lc.capture(name) : null;
+                    Var capturedRet = lc.capture(ret);
+                    Var capturedEvaluatedParams = lc.capture(evaluatedParams);
+                    Var capturedEvalContext = lc.capture(evalContext);
+                    @SuppressWarnings("unused")
+                    ParamVar result = lc.parameter("r", 0);
+                    ParamVar throwable = lc.parameter("t", 1);
+
+                    lc.body(accept -> {
+                        accept.ifElse(accept.isNull(throwable), success -> {
+                            // Check type parameters and return NO_RESULT if failed
+                            List<Param> evaluated = params.evaluated();
+                            LocalVar paramTypes = success.localVar("pt", success.newEmptyArray(Class.class, evaluated.size()));
+                            int idx = 0;
+                            for (Param p : evaluated) {
+                                success.set(paramTypes.elem(idx++), Const.of(classDescOf(p.type.name())));
```

Comment:
Not sure what `p.type` is, but if it is a Jandex `Type`, you can do just `classDescOf(p.type)`.